File: 04_Sleeping_Barber/main.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define CHAIRS 3
#define TOTAL_CUSTOMERS 8

static int waiting_customers = 0;
static int barber_sleeping = 0;
static int all_done = 0;

static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
static pthread_cond_t cond_barber = PTHREAD_COND_INITIALIZER;
static pthread_cond_t cond_customers = PTHREAD_COND_INITIALIZER;

static void *barber_thread(void *arg) {
    (void)arg;

    while (1) {
        pthread_mutex_lock(&mutex);

        while (waiting_customers == 0 && !all_done) {
            printf("Nema musterija. Berberin spava...\n");
            barber_sleeping = 1;
            pthread_cond_wait(&cond_barber, &mutex);
            barber_sleeping = 0;
        }

        if (waiting_customers == 0 && all_done) {
            pthread_mutex_unlock(&mutex);
            break;
        }

        /* Berberin preuzima jednu musteriju iz cekaonice */
        waiting_customers--;
        printf("Berberin poziva musteriju. (U cekaonici ostaje: %d)\n", waiting_customers);
        
        /* Budimo sisanje za musteriju */
        pthread_cond_signal(&cond_customers);
        pthread_mutex_unlock(&mutex);

        /* Simulacija sisanja van kriticne sekcije */
        printf("Berberin sisa musteriju...\n");
        usleep(40000);
        printf("Berberin je zavrsio sisanje.\n");
    }

    printf("Berberin zatvara radnju za danas.\n");
    return NULL;
}

static void *customer_thread(void *arg) {
    int id = *(int *)arg;
    free(arg);

    usleep((rand() % 150) * 1000);

    pthread_mutex_lock(&mutex);
    printf("Musterija %d stize u berbernicu.\n", id);

    if (waiting_customers < CHAIRS) {
        waiting_customers++;
        printf("Musterija %d seda u cekaonicu. (Ukupno ceka: %d)\n", id, waiting_customers);

        if (barber_sleeping) {
            printf("Musterija %d budi berberina!\n", id);
            pthread_cond_signal(&cond_barber);
        }

        /* Musterija ceka da je berberin prozove */
        pthread_cond_wait(&cond_customers, &mutex);
        printf("Musterija %d ulazi na stolicu za sisanje.\n", id);
    } else {
        printf("Cekaonica je puna! Musterija %d odlazi bez sisanja.\n", id);
    }

    pthread_mutex_unlock(&mutex);
    return NULL;
}

int main(void) {
    srand(time(NULL));
    pthread_t barber;
    pthread_t customers[TOTAL_CUSTOMERS];

    pthread_create(&barber, NULL, barber_thread, NULL);

    for (int i = 0; i < TOTAL_CUSTOMERS; i++) {
        int *id = malloc(sizeof(int));
        *id = i + 1;
        pthread_create(&customers[i], NULL, customer_thread, id);
    }

    for (int i = 0; i < TOTAL_CUSTOMERS; i++) {
        pthread_join(customers[i], NULL);
    }

    pthread_mutex_lock(&mutex);
    all_done = 1;
    pthread_cond_signal(&cond_barber);
    pthread_mutex_unlock(&mutex);

    pthread_join(barber, NULL);

    printf("\n[USPEH] Svi klijenti procesuirani, program uspesno zatvoren!\n");
    return 0;
}

File: 04_Sleeping_Barber/main.go
/*
 * Sleeping Barber Problem
 *
 * -----------------------------------------------------------
 * C vs Go:
 *   pthread_mutex_t + pthread_cond_t  ->  sync.Mutex + sync.Cond
 *   pthread_cond_wait(&c, &m)         ->  cond.Wait()  (otpusta mutex dok ceka)
 *   pthread_cond_signal(&c)           ->  cond.Signal()
 * -----------------------------------------------------------
 */

package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const (
	CHAIRS          = 3
	TOTAL_CUSTOMERS = 8
)

var (
	waitingCustomers int
	barberSleeping   bool
	allDone          bool

	mu            sync.Mutex
	condBarber    = sync.NewCond(&mu)
	condCustomers = sync.NewCond(&mu)
)

func barberThread(wg *sync.WaitGroup) {
	defer wg.Done()

	for {
		mu.Lock()

		for waitingCustomers == 0 && !allDone {
			fmt.Println("Nema musterija. Berberin spava...")
			barberSleeping = true
			condBarber.Wait()
			barberSleeping = false
		}

		if waitingCustomers == 0 && allDone {
			mu.Unlock()
			break
		}

		waitingCustomers--
		fmt.Printf("Berberin poziva musteriju. (U cekaonici ostaje: %d)\n", waitingCustomers)
		condCustomers.Signal()
		mu.Unlock()

		fmt.Println("Berberin sisa musteriju...")
		time.Sleep(40 * time.Millisecond)
		fmt.Println("Berberin je zavrsio sisanje.")
	}

	fmt.Println("Berberin zatvara radnju za danas.")
}

func customerThread(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	time.Sleep(time.Duration(rand.Intn(150)) * time.Millisecond)

	mu.Lock()
	fmt.Printf("Musterija %d stize u berbernicu.\n", id)

	if waitingCustomers < CHAIRS {
		waitingCustomers++
		fmt.Printf("Musterija %d seda u cekaonicu. (Ukupno ceka: %d)\n", id, waitingCustomers)

		if barberSleeping {
			fmt.Printf("Musterija %d budi berberina!\n", id)
			condBarber.Signal()
		}

		condCustomers.Wait()
		fmt.Printf("Musterija %d ulazi na stolicu za sisanje.\n", id)
	} else {
		fmt.Printf("Cekaonica je puna! Musterija %d odlazi bez sisanja.\n", id)
	}

	mu.Unlock()
}

func main() {
	var wg sync.WaitGroup

	wg.Add(1)
	go barberThread(&wg)

	for i := 1; i <= TOTAL_CUSTOMERS; i++ {
		wg.Add(1)
		go customerThread(i, &wg)
	}

	// Cekamo sve musterije
	// Napomena: moramo posebno da pratimo kraj musterija
	customerWg := sync.WaitGroup{}
	for i := 1; i <= TOTAL_CUSTOMERS; i++ {
		customerWg.Add(1)
		go func(id int) {
			defer customerWg.Done()
		}(i)
	}

	time.Sleep(3 * time.Second) // Cekamo da sve goroutine zavrse

	mu.Lock()
	allDone = true
	condBarber.Signal()
	mu.Unlock()

	wg.Wait()
	fmt.Println("\n[USPEH] Svi klijenti procesuirani, program uspesno zatvoren!")
}
